### ISSUE TITLE
libroadrunner==2.2.1 results in segmentation faults when using with ray (multiprocessing) and LLJIT

### ISSUE THREAD

**Issue by matthiaskoenig** [sys-bio/roadrunner#999]
Hi all,

still trying to find some solution for the segmentation faults I am seeing in #963
I am more and more sure that the latest versions of roadrunner do not work anymore correctly in python multiprocessing environments. I.e. if multiple cores read/write states and read models at the same time there seem to be issues. Really hard to pinpoint what is going on. 

Finally I could create an example which shows part of the issue (but the #963 seems to be bigger/more complex). The example below shows how when running simulations on multiple cores the LLJIT results in segmentation faults (when deactiving the LLJIT this works).
There seem to be bugs in the saveState/loadState introduced after version 2.1.3 which make this not work in a multprocessing/multicore environment any more. I am pretty sure these bugs are also the reason for #963 (because these all appeared after libroadrunner version 2.1.3) and make all versions after 2.1.3 unusable for me at the moment :/

Here the full example (below the errors).
[example.zip](https://github.com/sys-bio/roadrunner/files/8665212/example.zip)

```python
"""Example parallel simulation with ray.

requirements
    roadrunner==2.2.1
    ray
    pandas
"""
import time
from pathlib import Path

import pandas as pd
import ray

import roadrunner
from roadrunner import Config
Config.setValue(Config.LLVM_BACKEND, Config.LLJIT)


# start ray
ray.init(ignore_reinit_error=True)


@ray.remote
class SimulatorActorPath(object):
    """Ray actor to execute simulations."""

    def __init__(self, r: roadrunner.RoadRunner):
        self.r: roadrunner.RoadRunner = r

    def simulate(self, size=1):
        """Simulate."""
        print("Start simulations")
        ts = time.time()
        results = []
        for _ in range(size):
            self.r.resetAll()
            s = self.r.simulate(0, 100, steps=100)
            # create numpy array (which can be pickled), better use shared memory
            df = pd.DataFrame(s, columns=s.colnames)
            results.append(df)
        te = time.time()
        print("Finished '{}' simulations: {:2.2f} ms".format(size, (te - ts) * 1000))
        return results


def ray_example():
    """Ray example."""
    actor_count: int = 2  # cores to run this on

    model_path = Path(__file__).parent / "icg_body_flat.xml"
    rr: roadrunner.RoadRunner = roadrunner.RoadRunner(str(model_path))

    simulators = [SimulatorActorPath.remote(rr) for _ in range(actor_count)]

    # run simulations
    sim_per_actor = 10
    tc_ids = []
    for simulator in simulators:
        tcs_id = simulator.simulate.remote(size=sim_per_actor)
        tc_ids.append(tcs_id)

    results = ray.get(tc_ids)
    return results


if __name__ == "__main__":
    ray_example()
```
Results in

```
2022-05-10 23:50:57,748 INFO services.py:1412 -- View the Ray dashboard at http://127.0.0.1:8265
(SimulatorActorPath pid=1363828) Start simulations
(SimulatorActorPath pid=1363828) Finished '10' simulations: 13.58 ms
(SimulatorActorPath pid=1363835) *** SIGSEGV received at time=1652219460 on cpu 28 ***
(SimulatorActorPath pid=1363835) PC: @     0x7fcc8b429fd7  (unknown)  (unknown)
(SimulatorActorPath pid=1363835)     @     0x7fcc483080c0  (unknown)  (unknown)
(SimulatorActorPath pid=1363835)     @     0x7fca8d18f988         48  rrllvm::LLVMExecutableModel::getGlobalParameterValues()
(SimulatorActorPath pid=1363835)     @     0x7fca8d163b07         96  rr::ForwardSensitivitySolver::getModelParametersAsVector()
(SimulatorActorPath pid=1363835)     @     0x7fca8d15f5bf        144  rr::ForwardSensitivitySolver::constructorOperations()
(SimulatorActorPath pid=1363835)     @     0x7fca8d1629c1         32  rr::ForwardSensitivitySolver::syncWithModel()
(SimulatorActorPath pid=1363835)     @     0x7fca8d102ad9        128  rr::RoadRunnerImpl::syncAllSolversWithModel()
(SimulatorActorPath pid=1363835)     @     0x7fca8d0f4180        784  rr::RoadRunner::loadStateS()
(SimulatorActorPath pid=1363835)     @     0x7fca8d03f5d7       1024  _wrap_RoadRunner_loadStateS
(SimulatorActorPath pid=1363835)     @           0x5f3989  (unknown)  PyCFunction_Call
(SimulatorActorPath pid=1363835)     @           0x903b00  (unknown)  (unknown)
(SimulatorActorPath pid=1363835) [2022-05-10 23:51:00,089 E 1363835 1363835] logging.cc:321: *** SIGSEGV received at time=1652219460 on cpu 28 ***
(SimulatorActorPath pid=1363835) [2022-05-10 23:51:00,091 E 1363835 1363835] logging.cc:321: PC: @     0x7fcc8b429fd7  (unknown)  (unknown)
(SimulatorActorPath pid=1363835) [2022-05-10 23:51:00,091 E 1363835 1363835] logging.cc:321:     @     0x7fcc483080c0  (unknown)  (unknown)
(SimulatorActorPath pid=1363835) [2022-05-10 23:51:00,091 E 1363835 1363835] logging.cc:321:     @     0x7fca8d18f988         48  rrllvm::LLVMExecutableModel::getGlobalParameterValues()
(SimulatorActorPath pid=1363835) [2022-05-10 23:51:00,091 E 1363835 1363835] logging.cc:321:     @     0x7fca8d163b07         96  rr::ForwardSensitivitySolver::getModelParametersAsVector()
(SimulatorActorPath pid=1363835) [2022-05-10 23:51:00,091 E 1363835 1363835] logging.cc:321:     @     0x7fca8d15f5bf        144  rr::ForwardSensitivitySolver::constructorOperations()
(SimulatorActorPath pid=1363835) [2022-05-10 23:51:00,091 E 1363835 1363835] logging.cc:321:     @     0x7fca8d1629c1         32  rr::ForwardSensitivitySolver::syncWithModel()
(SimulatorActorPath pid=1363835) [2022-05-10 23:51:00,092 E 1363835 1363835] logging.cc:321:     @     0x7fca8d102ad9        128  rr::RoadRunnerImpl::syncAllSolversWithModel()
(SimulatorActorPath pid=1363835) [2022-05-10 23:51:00,092 E 1363835 1363835] logging.cc:321:     @     0x7fca8d0f4180        784  rr::RoadRunner::loadStateS()
(SimulatorActorPath pid=1363835) [2022-05-10 23:51:00,092 E 1363835 1363835] logging.cc:321:     @     0x7fca8d03f5d7       1024  _wrap_RoadRunner_loadStateS
(SimulatorActorPath pid=1363835) [2022-05-10 23:51:00,092 E 1363835 1363835] logging.cc:321:     @           0x5f3989  (unknown)  PyCFunction_Call
(SimulatorActorPath pid=1363835) [2022-05-10 23:51:00,092 E 1363835 1363835] logging.cc:321:     @           0x903b00  (unknown)  (unknown)
(SimulatorActorPath pid=1363835) Fatal Python error: Segmentation fault
(SimulatorActorPath pid=1363835) 
(SimulatorActorPath pid=1363835) Stack (most recent call first):
(SimulatorActorPath pid=1363835)   File "/home/mkoenig/envs/sbmlsim/lib/python3.8/site-packages/roadrunner/roadrunner.py", line 3024 in loadStateS
(SimulatorActorPath pid=1363835)   File "/home/mkoenig/envs/sbmlsim/lib/python3.8/site-packages/roadrunner/roadrunner.py", line 4397 in __setstate__
(SimulatorActorPath pid=1363835)   File "/home/mkoenig/envs/sbmlsim/lib/python3.8/site-packages/ray/serialization.py", line 164 in _deserialize_pickle5_data
(SimulatorActorPath pid=1363835)   File "/home/mkoenig/envs/sbmlsim/lib/python3.8/site-packages/ray/serialization.py", line 174 in _deserialize_msgpack_data
(SimulatorActorPath pid=1363835)   File "/home/mkoenig/envs/sbmlsim/lib/python3.8/site-packages/ray/serialization.py", line 215 in _deserialize_object
(SimulatorActorPath pid=1363835)   File "/home/mkoenig/envs/sbmlsim/lib/python3.8/site-packages/ray/serialization.py", line 309 in deserialize_objects
(SimulatorActorPath pid=1363835)   File "/home/mkoenig/envs/sbmlsim/lib/python3.8/site-packages/ray/worker.py", line 324 in deserialize_objects
(SimulatorActorPath pid=1363835)   File "/home/mkoenig/envs/sbmlsim/lib/python3.8/site-packages/ray/worker.py", line 430 in main_loop
(SimulatorActorPath pid=1363835)   File "/home/mkoenig/envs/sbmlsim/lib/python3.8/site-packages/ray/workers/default_worker.py", line 224 in <module>
2022-05-10 23:51:00,298 WARNING worker.py:1326 -- A worker died or was killed while executing a task by an unexpected system error. To troubleshoot the problem, check the logs for the dead worker. RayTask ID: ffffffffffffffff2498a8b3414477aa1d7967bb01000000 Worker ID: 62570256db496dde0a1d4b1a93b9d7222ca285d00d35988fc59bdb98 Node ID: f6812d4b7350f7f42acb47928d36884051a37cd76780518c6e77482c Worker IP address: 192.168.1.105 Worker port: 36985 Worker PID: 1363835
Traceback (most recent call last):
  File "/home/mkoenig/git/sbmlsim/src/sbmlsim/examples/multiprocessing/ray_example.py", line 67, in <module>
    ray_example()
  File "/home/mkoenig/git/sbmlsim/src/sbmlsim/examples/multiprocessing/ray_example.py", line 62, in ray_example
    results = ray.get(tc_ids)
  File "/home/mkoenig/envs/sbmlsim/lib/python3.8/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/home/mkoenig/envs/sbmlsim/lib/python3.8/site-packages/ray/worker.py", line 1765, in get
    raise value
ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
        class_name: SimulatorActorPath
        actor_id: 2498a8b3414477aa1d7967bb01000000
        pid: 1363835
        namespace: 2c992676-ed04-4044-b673-0bb4a4411381
        ip: 192.168.1.105
The actor is dead because its worker process has died. Worker exit type: SYSTEM_ERROR_EXIT

Process finished with exit code 1
```

**Comment by luciansmith** [comment]
For me, running your code on my ubuntu VM produces no errors, and running your code on Windows produces different errors.  For the Windows version I get something like this:

```
(pid=19072) Windows fatal exception: access violation
(pid=19072) 
(pid=8572) Windows fatal exception: access violation
(pid=8572) 
 pid=19072) Start simulations
 pid=19072) Finished '10' simulations: 6.85 ms
 pid=8572) Start simulations
 pid=8572) Finished '10' simulations: 5.85 ms
```

However, I can strip down your example to a much much smaller example and get the same errors:

```python
import ray
import roadrunner

# start ray
ray.init(ignore_reinit_error=True)

modelstr = """<?xml version="1.0" encoding="UTF-8"?>
<sbml xmlns="http://www.sbml.org/sbml/level3/version1/core" level="3" version="1">
  <model metaid="__main" id="__main">
    <listOfParameters>
      <parameter id="a" value="3" constant="true"/>
    </listOfParameters>
  </model>
</sbml>
"""

@ray.remote
class SimulatorActorPath(object):
    """Ray actor to execute simulations."""

    def __init__(self):
        self.r = roadrunner.RoadRunner(modelstr)

    def simulate(self):
        """Simulate."""
        print("Just read the value of 'a'")
        print(self.r.getValue("a"), "\n")

def ray_example():
    """Ray example."""
    actor_count: int = 1  # cores to run this on

    simulators = [SimulatorActorPath.remote() for _ in range(actor_count)]

    # run simulations
    # tc_ids = []
    for simulator in simulators:
        simulator.simulate.remote()

if __name__ == "__main__":
    ray_example()
```
gives me:
```
(pid=16232) Windows fatal exception: access violation
(pid=16232) 
 pid=16232) Just read the value of 'a'
 pid=16232) 3.0 ath
 pid=16232) torPath
```

As you can see, I get the errors even when only using one core, and I also get the same errors with earlier versions of libroadrunner.

So, while I am getting some sort of error on Windows, it's not the same as yours, and seems to be some sort of problem using ray instead.  And when I run it on ubuntu, everything seems to work fine.

(The bug above is particularly mysterious:  there seems to be some sort of residual string left over?  From the class *name*, somehow?)

Just to be completist, I also ran the smaller ray example on Ubuntu, where it produced no errors and no output at all (not even the prints).

I can also report that unrelatedly, I've been using 'concurrent' to run roadrunner in parallel and have not had problems there, either, neither on Windows nor Ubuntu.

I'm not sure this helps, but I at least am interested to see if my cut down version of your example runs on your system.

**Comment by luciansmith** [comment]
Further update:  the Windows errors only appear with Python 3.8.  Switching to 3.9.1 results in zero errors for both your example and the reduced example.  

Ubuntu was using 3.8.10, and saw no errors.

I thought perhaps the problem might be that I was using 3.8.3, but upgrading to 3.8.10 (on Windows) gave me the same errors.

I know ray isn't out yet for 3.10, so perhaps you're still using 3.8?  It might be worth seeing if switching to 3.9 helps.

**Comment by matthiaskoenig** [comment]
Hi Lucian,

using python 3.9 already. I did not see any errors with your example.
As you see this is very hard too debug. 
I have the feeling that the models have to be larger for the issues to happen, but this is pretty reproducible in my lab (i.e. a handful of students on different linux machines get these segmentation faults).
I wanted to drop the ray part tomorrow and see if I can get things working with naive multiprocessing. 

Thanks for looking into this. 
Best Matthias

**Comment by luciansmith** [comment]
Another thing you might try:  seeding your parallel objects with a save state string instead of the roadrunner object:

```python
@ray.remote
class SimulatorActorPath(object):
    """Ray actor to execute simulations."""

    def __init__(self, savestr: str):
        self.r = roadrunner.RoadRunner()
        self.r.loadStateS(savestr)

    def simulate(self, size=1):
        """Simulate."""
        print("Start simulations")
        ts = time.time()
        results = []
        for _ in range(size):
            self.r.resetAll()
            s = self.r.simulate(0, 100, steps=100)
            # create numpy array (which can be pickled), better use shared memory
            df = pd.DataFrame(s, columns=s.colnames)
            results.append(df)
        te = time.time()
        print("Finished '{}' simulations: {:2.2f} ms".format(size, (te - ts) * 1000))
        return results


def ray_example():
    """Ray example."""
    actor_count: int = 2  # cores to run this on

    model_path = Path(__file__).parent / "icg_body_flat.xml"
    rr: roadrunner.RoadRunner = roadrunner.RoadRunner(str(model_path))
    savestr = rr.saveStateS()

    simulators = [SimulatorActorPath.remote(savestr) for _ in range(actor_count)]

    # run simulations
    sim_per_actor = 10
    tc_ids = []
    for simulator in simulators:
        tcs_id = simulator.simulate.remote(size=sim_per_actor)
        tc_ids.append(tcs_id)

    results = ray.get(tc_ids)
    return results
```

**Comment by matthiaskoenig** [comment]
Duplicate of #963 Same issue. Closing this.
